packages/xdl/src/Project.ts

```diff
@@ -900,135 +921,235 @@ export async function publishAsync(
+      projectRoot,
+      'ios.publishSourceMapPath',
+      exp.ios.publishSourceMapPath,
+      iosSourceMap
+    );
+  }
+
+  if (exp.ios && exp.ios.publishManifestPath) {
+    await _writeArtifactSafelyAsync(
+      projectRoot,
+      'ios.publishManifestPath',
+      exp.ios.publishManifestPath,
+      JSON.stringify(iosManifest)
+    );
+    const context = StandaloneContext.createUserContext(projectRoot, exp);
+    const { supportingDirectory } = IosWorkspace.getPaths(context);
+    if (releaseChannel && fs.existsSync(path.join(supportingDirectory, 'EXShell.plist'))) {
+      // This is an ExpoKit app, set properties in EXShell.plist
+      await IosPlist.modifyAsync(supportingDirectory, 'EXShell', (shellPlist: any) => {
+        shellPlist.releaseChannel = releaseChannel;
+        return shellPlist;
+      });
     }
+    if (fs.existsSync(path.join(supportingDirectory, 'Expo.plist'))) {
+      // This is an app with expo-updates installed, set properties in Expo.plist
+      await IosPlist.modifyAsync(supportingDirectory, 'Expo', (configPlist: any) => {
+        configPlist.EXUpdatesURL = iosManifestUrl;
+        configPlist.EXUpdatesSDKVersion = exp.sdkVersion;
+        if (releaseChannel) {
+          configPlist.EXUpdatesReleaseChannel = releaseChannel;
+        }
+        return configPlist;
+      });
+    }
+  }
 
-    if (exp.android && exp.android.publishManifestPath) {
-      await _writeArtifactSafelyAsync(
-        projectRoot,
-        'android.publishManifestPath',
-        exp.android.publishManifestPath,
-        JSON.stringify(androidManifest)
+  if (exp.android && exp.android.publishManifestPath) {
+    await _writeArtifactSafelyAsync(
+      projectRoot,
+      'android.publishManifestPath',
+      exp.android.publishManifestPath,
+      JSON.stringify(androidManifest)
+    );
+
+    let constantsPath = path.join(
+      projectRoot,
+      'android',
+      'app',
+      'src',
+      'main',
+      'java',
+      'host',
+      'exp',
+      'exponent',
+      'generated',
+      'AppConstants.java'
+    );
+    if (fs.existsSync(constantsPath)) {
+      // This is an ExpoKit app
+      // We need to add EmbeddedResponse instances on Android to tell the runtime
+      // that the shell app manifest and bundle is packaged.
+      await ExponentTools.deleteLinesInFileAsync(
+        `START EMBEDDED RESPONSES`,
+        `END EMBEDDED RESPONSES`,
+        constantsPath
+      );
+      await ExponentTools.regexFileAsync(
+        '// ADD EMBEDDED RESPONSES HERE',
+        `
+        // ADD EMBEDDED RESPONSES HERE
+        // START EMBEDDED RESPONSES
+        embeddedResponses.add(new Constants.EmbeddedResponse("${androidManifestUrl}", "assets://shell-app-manifest.json", "application/json"));
+        embeddedResponses.add(new Constants.EmbeddedResponse("${androidManifest.bundleUrl}", "assets://shell-app.bundle", "application/javascript"));
+        // END EMBEDDED RESPONSES`,
+        constantsPath
```

Comment:
move all this and the rest of the code below it in the `if (exp.android && exp.android.publishManifestPath)` clause into a subroutine (github wont let me highlight the entire code chunk)